cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,590 @@
+  for (int i = 0; i < input.num_children(); ++i) {
+    auto child = out->children[i];
+    auto col   = input.child(i);
+    NANOARROW_RETURN_NOT_OK(ArrowSchemaSetName(child, metadata.name.c_str()));
+    child->flags = col.nullable() ? ARROW_FLAG_NULLABLE : 0;
+
+    if (col.type().id() == cudf::type_id::EMPTY) {
+      NANOARROW_RETURN_NOT_OK(ArrowSchemaSetType(child, NANOARROW_TYPE_NA));
+      continue;
+    }
+
+    NANOARROW_RETURN_NOT_OK(cudf::type_dispatcher(col.type(),
+                                                  detail::dispatch_to_arrow_type{},
+                                                  col,
+                                                  col.type().id(),
+                                                  metadata.children_meta[i],
+                                                  child));
+  }
+
+  return NANOARROW_OK;
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::list_view>(column_view input,
+                                                        cudf::type_id,
+                                                        column_metadata const& metadata,
+                                                        struct ArrowSchema* out)
+{
+  NANOARROW_RETURN_NOT_OK(ArrowSchemaSetType(out, NANOARROW_TYPE_LIST));
+  auto child = input.child(0);
+  if (child.type().id() == cudf::type_id::EMPTY) {
+    return ArrowSchemaSetType(out->children[0], NANOARROW_TYPE_NA);
+  }
+  out->flags = child.nullable() ? ARROW_FLAG_NULLABLE : 0;
+
+  auto child_meta =
+    metadata.children_meta.empty() ? column_metadata{"element"} : metadata.children_meta[0];
+  return cudf::type_dispatcher(child.type(),
+                               detail::dispatch_to_arrow_type{},
+                               child,
+                               child.type().id(),
+                               child_meta,
+                               out->children[0]);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::dictionary32>(column_view input,
+                                                           cudf::type_id,
+                                                           column_metadata const& metadata,
+                                                           struct ArrowSchema* out)
+{
+  cudf::dictionary_column_view dview{input};
+  NANOARROW_RETURN_NOT_OK(ArrowSchemaAllocateDictionary(out));
+  NANOARROW_RETURN_NOT_OK(ArrowSchemaSetType(out, id_to_arrow_type(dview.indices().type().id())));
+
+  auto dict_keys = dview.keys();
+  return cudf::type_dispatcher(
+    dict_keys.type(),
+    detail::dispatch_to_arrow_type{},
+    dict_keys,
+    dict_keys.type().id(),
+    metadata.children_meta.empty() ? column_metadata{"keys"} : metadata.children_meta[0],
+    out->dictionary);
+}
+
+template <typename T>
+static void device_buffer_finalize(ArrowBufferAllocator* allocator, uint8_t*, int64_t)
+{
+  auto* unique_buffer = reinterpret_cast<std::unique_ptr<T>*>(allocator->private_data);
+  delete unique_buffer;
+}
```

Comment:
> So at the end of set_buffer ownership has changed hands into a unique_ptr whose scope is managed (manually) by arrow's allocator. Is that right? And I guess that means private_data really can be used to hold anything and that is an intentional part of the design, we're not just abusing the fact that it's a void * to stuff something else in there?

That's exactly the case :smile:

Reply:
> Why does it need to be static? Why is the scope/linkage of this function special?

It doesn't *need* to be static, it's just convenient to ensure that it isn't being exposed outside this file and allow more aggressive optimizations by the compiler if it desires to do so as we only use this as a function pointer for release callbacks. If you'd prefer it to not be marked static, that's perfectly fine.

Reply:
As per the direction on [the other thread](https://github.com/rapidsai/cudf/pull/15047#discussion_r1508108642) I think we can remove the static since it's redundant in the anonymous namespace.